warp/backend.go

```diff
@@ -174,3 +197,39 @@ func (b *backend) GetMessage(messageID ids.ID) (*avalancheWarp.UnsignedMessage,
 
 	return unsignedMessage, nil
 }
+
+// addOffChainMessage parses the message and payload and adds to off-chain map if valid
+func (b *backend) addOffChainMessage(message hexutil.Bytes) error {
+	unsignedMessage, err := avalancheWarp.ParseUnsignedMessage(message)
+	if err != nil {
+		return fmt.Errorf("%w: %s", warp.ErrInvalidWarpMsg, err)
+	}
+
+	id, err := parsePayload(unsignedMessage)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Adding warp message to off-chain message map", "messageID", unsignedMessage.ID(), "decrypted warp message", hex.EncodeToString(unsignedMessage.Bytes()))
+	b.offChainMessageMap[id] = unsignedMessage
+	return nil
+}
+
+// parsePayload parses the payload of an unsigned message and returns the ID of the valid payload.
+// If the payload is a valid block hash, it returns the block ID; otherwise, it returns the unsigned message ID.
+// It returns ErrInvalidWarpMsgPayload on an invalid payload.
+func parsePayload(unsignedMessage *avalancheWarp.UnsignedMessage) (ids.ID, error) {
+	id := unsignedMessage.ID()
+
+	blockHashPayload, err := payload.ParseHash(unsignedMessage.Payload)
+	if err == nil {
+		id = blockHashPayload.Hash
+	} else {
+		_, err = payload.Parse(unsignedMessage.Payload)
+		if err != nil {
+			return ids.Empty, fmt.Errorf("%w: %s", warp.ErrInvalidWarpMsgPayload, err)
+		}
+	}
+
+	return id, nil
+}
```

Comment:
Can we use `payload.Parse` directly and then have a type switch on the returned value to check which off chain messages map it needs to be placed in?

This should be pretty short and is only used once, so I think it would make sense to inline as well.

Reply:
yup